### ISSUE TITLE
Child gets recomposed more than once at startup in Compose Desktop

### ISSUE THREAD

**Issue by racka98** [arkivanov/Decompose#56]
I have 2 problems at the moment.

1. When you launch the app on Compose Desktop the initial configuration destination gets called more than once which causes the Children block to be called more than once causing multiple recompositions which can lead to a crash.

You can see here on the first screenshot that at the launch there are multiple calls to same thing (related to the Home Screen):
<img width="404" alt="Screenshot 2022-03-23 090508" src="https://user-images.githubusercontent.com/41214249/159636275-62003539-36dd-4348-8363-0a72cde9fd91.png">

But after you navigate to something like a Settings Screen. You only see one call for each. Even when you pop back to the Home screen the home screen composable will be called only once this time:
<img width="379" alt="Screenshot 2022-03-23 090709" src="https://user-images.githubusercontent.com/41214249/159636578-99af8b80-244a-41c2-af05-6f0cbc2e3e08.png">

I am using Decompose version: 0.5.2. Github link to the navigation and ui part of the app in question is this: https://github.com/Thinkrchive/Thinkrchive-Multiplatform/tree/main/desktopApp/src/jvmMain/kotlin/work/racka/thinkrchive/v2/desktop/ui

The navigation folder contains RootComponent and it's implementations.
The screen folder contains each screen composable with it component Interface and Implementation.

I followed the official guide for using compose-extensions.

2. Navigating to my About screen fails.
When I click the About button the event is not propagated from the comoposable to the component creation inside my [NavHostComponent](https://github.com/Thinkrchive/Thinkrchive-Multiplatform/blob/dbab7c9eca472824ca2f49677041d039cfa0b0ab/desktopApp/src/jvmMain/kotlin/work/racka/thinkrchive/v2/desktop/ui/navigation/NavHostComponent.kt#L64)

What's weird is that every other call like for settingsScreen or donateScreen work as expected just the aboutScreen call that has issues.

Also using crossfade() or crossfadeScale() crashes the app or refuses any navigation calls.

**Comment by arkivanov** [comment]
Hello. I didn't deep dive into the issue yet. But from the first sight, we should never rely on how much time recompositions should happen.

Ideally components and their contexts should be created outside of Composables functions. E.g. the Root component should be created in the main function, before Compose `application` is called. See the example: https://github.com/JetBrains/compose-jb/blob/f1854a9dce4c38fce6101f0844a96ebd75ca2ad6/examples/todoapp/desktop/src/jvmMain/kotlin/example/todo/desktop/Main.kt#L30

If this is not possible for some reason, then there is `remember` function which remembers values between compositions.

The navigation should be always performed as side effects, never from the composition calls.

**Comment by arkivanov** [comment]
I believe this should be mentioned in the documentation, I will add. Thanks for raising.

**Comment by racka98** [comment]
> Hello. I didn't deep dive into the issue yet. But from the first sight, we should never rely on how much time recompositions should happen.
> 
> Ideally components and their contexts should be created outside of Composables functions. E.g. the Root component should be created in the main function, before Compose `application` is called. See the example: [JetBrains/compose-jb@`f1854a9`/examples/todoapp/desktop/src/jvmMain/kotlin/example/todo/desktop/Main.kt#L30](https://github.com/JetBrains/compose-jb/blob/f1854a9dce4c38fce6101f0844a96ebd75ca2ad6/examples/todoapp/desktop/src/jvmMain/kotlin/example/todo/desktop/Main.kt#L30)
> 
> If this is not possible for some reason, then there is `remember` function which remembers values between compositions.
> 
> The navigation should be always performed as side effects, never from the composition calls.

This method seems to have stopped the crashes which were caused by my ViewModel being created more than once rapidly and causing SQLDelight to crash my app due to conflicting transactions.  
Even though the first created Child is still being called more than once it no longer crashes or freezes the app like before. 
Thanks for pointing that out.

But the navigation to About screen issue still persists. I don't know why my onAboutClicked() event is not propagating to the respective component inside RootComponent [implementation](https://github.com/Thinkrchive/Thinkrchive-Multiplatform/blob/dbab7c9eca472824ca2f49677041d039cfa0b0ab/desktopApp/src/jvmMain/kotlin/work/racka/thinkrchive/v2/desktop/ui/navigation/NavHostComponent.kt#L64) despite other similar events being received. Is there something else I am missing?


https://user-images.githubusercontent.com/41214249/159673844-0b702d09-7051-4c94-9e08-2207392b5d81.mp4

**Comment by arkivanov** [comment]
Do you want to apply my suggestions? Components and contexts should not be created directly in composable functions. Either in the main app or in `remember`.

**Comment by racka98** [comment]
I understand what you mean. All of my components are created outside of the composables just like in the documentations but my current issue is with the single click (onAboutClick) for navigation to one route (About Screen) that does not activate no matter what I try. Maybe I should open a new issue since it's not directly to recomposition. Thank you for your time.

I'll close this issue since I've already moved component creation outiside of the `application` compose function. Sorry for the confusion.

**Comment by arkivanov** [comment]
Let's keep this issue open for a while, I want to add this to the documentation.

As far I see by the link provided above, components are created inside Composable functions. E.g. in the [App function](https://github.com/Thinkrchive/Thinkrchive-Multiplatform/blob/dbab7c9eca472824ca2f49677041d039cfa0b0ab/desktopApp/src/jvmMain/kotlin/work/racka/thinkrchive/v2/desktop/ui/navigation/App.kt#L21).

**Comment by racka98** [comment]
Already fixed that as per your recommendation. It's already moved to the main function that's why it not longer crashes like before.

See [Main.kt](https://github.com/Thinkrchive/Thinkrchive-Multiplatform/blob/e1f58226194822897f6efe8fd54eab98497afb5d/desktopApp/src/jvmMain/kotlin/work/racka/thinkrchive/v2/desktop/Main.kt#L14) in the new commit

**Comment by arkivanov** [comment]
Thanks, I will try to check the project today or tomorrow.

**Comment by arkivanov** [comment]
@racka98 It seems like you forgot to add parentheses [here](https://github.com/Thinkrchive/Thinkrchive-Multiplatform/blob/e1f58226194822897f6efe8fd54eab98497afb5d/desktopApp/src/jvmMain/kotlin/work/racka/thinkrchive/v2/desktop/ui/screens/splitpane/HomePaneComponentImpl.kt#L41). It should be `onAboutClicked()` not just `onAboutClicked`.

**Comment by racka98** [comment]
Oh wow. Don't know how I missed that. Thanks for your timel
